clippy_utils/src/lib.rs

```diff
@@ -3492,3 +3492,88 @@ pub fn is_receiver_of_method_call(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool
     }
     false
 }
+
+pub fn expr_requires_coercion<'tcx>(cx: &LateContext<'tcx>, expr: &Expr<'tcx>) -> bool {
```

Comment:
Would be great to have a doc comment here since it's pretty subtle. Something along the lines of `expr` either has a coercion or propagates a coercion from one of its sub expressions

Some guidance would be good too, i.e. that you should use this when you want to avoid suggesting replacing an expression that is currently a coercion site or coercion propagating expression with one that is not